vsintegration/src/FSharp.Editor/Formatting/EditorFormattingService.fs

```diff
@@ -68,6 +73,79 @@ type internal FSharpEditorFormattingService
+            do! Option.guard (startLineIdx <> sourceText.Lines.IndexOf span.End)
+            
+            let startLine = sourceText.Lines.[startLineIdx]
+
+            // VS quirk: if we're pasting on an empty line which has automatically been
+            // indented (i.e. by ISynchronousIndentationService), then the pasted span
+            // includes this automatic indentation. When pasting, we only care about what
+            // was actually in the clipboard.
+            let fixedSpan =
+                let pasteText = sourceText.GetSubText(span)
+                let pasteTextString = pasteText.ToString()
+
+                if currentClipboard.Length > 0 && pasteTextString.EndsWith currentClipboard then
+                    let prepended = pasteTextString.[0..pasteTextString.Length-currentClipboard.Length-1]
+                    
+                    // Only strip off leading indentation if the pasted span is otherwise
+                    // identical to the clipboard (ignoring leading spaces).
+                    if prepended |> Seq.forall ((=) ' ') then
+                        TextSpan(span.Start + prepended.Length, span.Length - prepended.Length)
+                    else
+                        span
+                else
+                    span
+
+            // Calculate the indentation of the line we pasted onto
+            let currentIndent =
+                let priorStartSpan = TextSpan(startLine.Span.Start, startLine.Span.Length - (startLine.Span.End - fixedSpan.Start))
+                
+                sourceText.GetSubText(priorStartSpan).ToString()
+                |> Seq.takeWhile ((=) ' ')
+                |> Seq.length
+
+            let fixedPasteText = sourceText.GetSubText(fixedSpan)
+            let leadingIndentation = fixedPasteText.ToString() |> getIndentation
+                
+            let stripIndentation charsToRemove =
+                let searchIndent = String.replicate charsToRemove " "
+                let newText = String.replicate currentIndent " "
+            
+                fixedPasteText.Lines
+                |> Seq.indexed
+                |> Seq.choose (fun (i, line) ->
+                    if line.ToString().StartsWith searchIndent then
+                        TextChange(TextSpan(line.Start + fixedSpan.Start, charsToRemove), if i = 0 then "" else newText)
+                        |> Some
+                    else
+                        None
+                )
+            
+            if leadingIndentation > 0 then
+                return stripIndentation leadingIndentation
+            else
+                let nextLineShouldBeIndented = FSharpIndentationService.IndentShouldFollow(documentId, sourceText, filePath, span.Start, parsingOptions)
+                
+                let removeIndentation =
+                    let nextLineIndent = fixedPasteText.Lines.[1].ToString() |> getIndentation
```

Comment:
Is `.[1]` is safe here?

Reply:
There’s a check above that checks that the span is multi line

Reply:
It seems difficult to see that check. Can we explicitly guard right before it for sanity?